ffi/derive/src/ffi_fn.rs

```diff
@@ -189,13 +195,19 @@ fn gen_output_assignment_stmts(fn_descriptor: &FnDescriptor) -> TokenStream {
             let src_type = receiver.src_type();
 
             if matches!(src_type, syn::Type::Path(_)) {
-                return quote! {
-                    if __out_ptr.is_null() {
-                        return Err(iroha_ffi::FfiReturn::ArgIsNull);
+                match output_arg.src_type() {
+                    // NOTE: case like fn(mut self, ...) -> Self (builder pattern)
+                    syn::Type::Path(path) if path.path.is_ident("Self") => {
```

Comment:
I'm not sure, but this looks like you should be using `SelfResolver`. Also, what if the type is used as input argument?

Reply:
> Also, what if the type is used as input argument?

Outer `if` is responsible for handling such case (match `self` and `mut self`).

```rust
if matches!(src_type, syn::Type::Path(_))
```

This change is not strictly relevant for the adding support for trait impls, but i find out that we currently can't convert to ffi functions like `fn(self) -> T`, so this change fix this.

Reply:
you probably mean methods, functions cannot have self receiver. So what about [this](https://github.com/hyperledger/iroha/blob/iroha2-dev/ffi/tests/ffi_export.rs#L50) test or [this](https://github.com/hyperledger/iroha/blob/iroha2-dev/ffi/tests/ffi_export.rs#L57)?

Reply:
ah, I think I get you now. Please follow up with a test case for this

Reply:
yes, if self is not returned from the ffi function memory will be leaked because in the generated body of the function `core::ptr::read` is used. PM me if you need more detailed explanation

Reply:
> I'm not sure, but this looks like you should be using SelfResolver
 
i think you are right, current solution would fail if instead of `fn(self) -> Self` someone would write `fn(self) -> Type` where `Self == Type`.